test/screenshot/lib/cli-arg-parser.js

```diff
@@ -114,10 +147,167 @@ E.g., 'origin/master' (default), 'HEAD', 'feat/foo/bar', 'fad7ed3'.
     return this.args_['mdc_diff_base'];
   }
+
+  /**
+   * @param {string} rawDiffBase
+   * @param {string} defaultGoldenPath
+   * @return {!Promise<!DiffSource>}
+   */
+  async parseDiffBase({
+    rawDiffBase = this.diffBase,
+    defaultGoldenPath = this.goldenPath,
+  } = {}) {
+    await this.gitRepo_.fetch();
+
+    // Diff against a public `golden.json` URL.
+    // E.g.: `--mdc-diff-base=https://storage.googleapis.com/.../golden.json`
+    const isUrl = HTTP_URL_REGEX.test(rawDiffBase);
+    if (isUrl) {
+      return this.createPublicUrlDiffSource_(rawDiffBase);
+    }
+
+    // Diff against a local `golden.json` file.
+    // E.g.: `--mdc-diff-base=/tmp/golden.json`
+    const isLocalFile = await fs.exists(rawDiffBase);
+    if (isLocalFile) {
+      return this.createLocalFileDiffSource_(rawDiffBase);
+    }
+
+    const [inputGoldenRef, inputGoldenPath] = rawDiffBase.split(':');
+    const goldenFilePath = inputGoldenPath || defaultGoldenPath;
+    const fullGoldenRef = await this.gitRepo_.getFullSymbolicName(inputGoldenRef);
+
+    // Diff against a specific git commit.
+    // E.g.: `--mdc-diff-base=abcd1234`
+    if (!fullGoldenRef) {
+      return this.createCommitDiffSource_(inputGoldenRef, goldenFilePath);
+    }
+
+    const {remoteRef, localRef, tagRef} = this.getRefType_(fullGoldenRef);
+
+    // Diff against a remote git branch.
+    // E.g.: `--mdc-diff-base=origin/master` or `--mdc-diff-base=origin/feat/button/my-fancy-feature`
+    if (remoteRef) {
+      return this.createRemoteBranchDiffSource_(remoteRef, goldenFilePath);
+    }
+
+    // Diff against a remote git tag.
+    // E.g.: `--mdc-diff-base=v0.34.1`
+    if (tagRef) {
+      return this.createRemoteTagDiffSource_(tagRef, goldenFilePath);
+    }
+
+    // Diff against a local git branch.
+    // E.g.: `--mdc-diff-base=master` or `--mdc-diff-base=HEAD`
+    return this.createLocalBranchDiffSource_(localRef, goldenFilePath);
+  }
+
+  createPublicUrlDiffSource_(publicUrl) {
+    return {
+      publicUrl,
+      localFilePath: null,
+      gitRevision: null,
+    };
+  }
+
+  createLocalFileDiffSource_(localFilePath) {
+    return {
+      publicUrl: null,
+      localFilePath,
+      gitRevision: null,
+    };
+  }
+
+  createCommitDiffSource_(commit, snapshotFilePath) {
+    return {
+      publicUrl: null,
+      localFilePath: null,
+      gitRevision: {
+        commit,
+        snapshotFilePath,
+        remote: null,
+        branch: null,
+        tag: null,
+      },
+    };
+  }
+
+  getRefType_(fullRef) {
+    const getShortGoldenRef = (type) => {
+      const regex = new RegExp(`^refs/${type}s/(.+)$`);
+      const match = regex.exec(fullRef) || [];
+      return match[1];
+    };
+
+    const remoteRef = getShortGoldenRef('remote');
+    const localRef = getShortGoldenRef('head');
+    const tagRef = getShortGoldenRef('tag');
+
+    return {remoteRef, localRef, tagRef};
+  }
+
+  async createRemoteBranchDiffSource_(remoteRef, snapshotFilePath) {
```

Comment:
Add annotations for params and return types here.

Reply:
Done